src/fluree/db/query/optimize.cljc

```diff
@@ -36,3 +44,184 @@
+          default-selectivity)))))
+
+(defn split-by-optimization-boundaries
+  "Split where clause into segments separated by optimization boundaries.
+   Optimizable patterns (:tuple, :class, :id) are grouped together.
+   All other patterns (boundaries) separate the segments."
+  [where-clause]
+  (reduce
+   (fn [segments pattern]
+     (if (optimizable-pattern? pattern)
+       (if (and (seq segments)
+                (= :optimizable (:type (peek segments))))
+         (update segments (dec (count segments))
+                 update :data conj pattern)
+         (conj segments {:type :optimizable :data [pattern]}))
+       ;; Boundary - add as separate segment
+       (conj segments {:type :boundary :data pattern})))
+   []
+   where-clause))
+
+(defn optimize-segment
+  "Optimize a single segment by sorting patterns by selectivity"
+  [db stats patterns]
+  (if (or (nil? stats) (empty? stats))
+    ;; No stats, return as-is
+    patterns
+    ;; Sort by selectivity (lower = more selective = execute first)
+    (let [with-scores (mapv (fn [pattern]
+                              {:pattern pattern
+                               :score (or (calculate-selectivity db stats pattern)
+                                          default-selectivity)})
+                            patterns)
+          sorted      (sort-by :score with-scores)]
+      (mapv :pattern sorted))))
+
+(defn optimize-patterns
+  "Reorder patterns for optimal execution based on statistics.
+   Splits on optimization boundaries and optimizes each segment independently."
+  [db where-clause]
+  (let [stats (:stats db)]
+    (if (or (nil? stats) (empty? stats))
+      ;; No stats available, return as-is
+      where-clause
+      ;; Split into segments and optimize each
+      (let [segments (split-by-optimization-boundaries where-clause)]
+        (into []
+              (mapcat (fn [segment]
+                        (if (= :optimizable (:type segment))
+                          (optimize-segment db stats (:data segment))
+                          [(:data segment)])))
+              segments)))))
+
+(defprotocol Optimizable
+  "Protocol for query optimization based on database statistics."
+
+  (-reorder [db parsed-query]
+    "Reorder query patterns based on database statistics.
+
+    Returns a channel that will contain the optimized query with patterns
+    reordered for optimal execution. If the database has no statistics
+    available, returns the query unchanged.
+
+    Parameters:
+      db - The database (FlakeDB, AsyncDB, etc.)
+      parsed-query - The parsed query from fql/parse-query
+
+    Returns:
+      Channel containing optimized query")
+
+  (-explain [db parsed-query]
+    "Generate an execution plan for the query showing optimization details.
+
+    Returns a channel that will contain a query plan map with:
+    - :query - The optimized query
+    - :plan - Execution plan details including:
+      - :optimization - :reordered, :unchanged, or :none
```

Comment:
What's the difference between `:unchanged` and `:none`?

Reply:
It means optimization was run but there was no change, vs no statistics were available and no optimization was attempted.